deepmd/utils/tabulate.py

```diff
@@ -0,0 +1,252 @@
+from tensorflow.python.framework import tensor_util
+
+class DeepTabulate():
+    """
+    Class for tabulation.
+    Compress a model, which including tabulating the embedding-net. 
+    The table is composed of fifth-order polynomial coefficients and is assembled from two sub-tables. The first table takes the stride(parameter) as it\'s uniform stride, while the second table takes 10 * stride as it\s uniform stride 
+    The range of the first table is automatically detected by deepmd-kit, while the second table ranges from the first table\'s upper boundary(upper) to the extrapolate(parameter) * upper.
+    """
+    def __init__(self,
+                 model_file,
+                 data_type,
+                 type_one_side = False) -> None:
+        """
+        Constructor
+
+        Parameters
+        ----------
+        model_file
+                The frozen model
+        data_type
+                The precision of the table. Supported options are {1}
+        type_one_side
+                Try to build N_types tables. Otherwise, building N_types^2 tables
+        """
+
+        self.model_file = model_file
+        self.data_type = data_type
+        self.type_one_side = type_one_side
+
+        self.graph, self.graph_def = self._load_graph()
+        self.sess = tf.Session(graph = self.graph)
+
+        self.sub_graph, self.sub_graph_def = self._load_sub_graph()
+        self.sub_sess = tf.Session(graph = self.sub_graph)
+
+        self.sel_a = self.graph.get_operation_by_name('DescrptSeA').get_attr('sel_a')
+        self.ntypes = self._get_tensor_value(self.graph.get_tensor_by_name ('descrpt_attr/ntypes:0'))
+
+        self.davg = self._get_tensor_value(self.graph.get_tensor_by_name ('descrpt_attr/t_avg:0'))
+        self.dstd = self._get_tensor_value(self.graph.get_tensor_by_name ('descrpt_attr/t_std:0'))
+
+        self.filter_variable_nodes = self._load_matrix_node()
+        self.layer_size = int(len(self.filter_variable_nodes) / (self.ntypes * self.ntypes * 2))
+        self.table_size = self.ntypes * self.ntypes
+        if type_one_side :
+            self.layer_size = int(len(self.filter_variable_nodes) / (self.ntypes * 2))
+            self.table_size = self.ntypes
+        # self.value_type = self.filter_variable_nodes["filter_type_0/matrix_1_0"].dtype #"filter_type_0/matrix_1_0" must exit~
+        # get trained variables
+        self.bias = self._get_bias()
+        self.matrix = self._get_matrix()
+        # self.matrix_layer_3 must exist
+        # self.data_type = type(self.matrix["layer_1"][0][0][0])
+        assert self.matrix["layer_1"][0].size > 0, "no matrix exist in matrix array!"
+        self.last_layer_size = self.matrix["layer_" + str(self.layer_size)][0].shape[1]
+        # define tables
+        self.data = {}
+
+        # TODO: Need a check function to determine if the current model is properly
+
+    def build(self, 
+              min_nbor_dist,
+              rcut,
+              rcut_smth,
```

Comment:
`rcut` is recorded in the model file. `rcut_smth` is not but we should added it to the model file.